File: android/app/src/main/java/ai/strainease/app/ui/strain/StrainPersonalInsight.kt
package ai.strainease.app.ui.strain

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.QueryStats
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import ai.strainease.app.data.ReliefLog

/**
 * Patient-specific "works for X" chip on the strain detail page.
 *
 * Mirrors `src/components/saved/StrainPersonalInsight.tsx` from PR-W1:
 * shows once the patient has logged this strain at least 3 times,
 * tries saved conditions first, then falls back to any condition
 * present in logs. Stays silent when there isn't enough data.
 */
@Composable
fun StrainPersonalInsight(
    logs: List<ReliefLog>,
    savedConditions: List<String>,
    modifier: Modifier = Modifier,
) {
    val snapshot = remember(logs, savedConditions) {
        snapshot(logs, savedConditions)
    }
    if (snapshot != null) {
        Surface(
            modifier = modifier
                .fillMaxWidth()
                .border(
                    width = 1.dp,
                    color = MaterialTheme.colorScheme.primary.copy(alpha = 0.25f),
                    shape = RoundedCornerShape(50),
                ),
            color = MaterialTheme.colorScheme.primary.copy(alpha = 0.08f),
            shape = RoundedCornerShape(50),
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 12.dp, vertical = 10.dp),
            ) {
                Icon(
                    imageVector = Icons.Filled.QueryStats,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.primary,
                    modifier = Modifier.padding(end = 8.dp),
                )
                Text(
                    text = labelFor(snapshot),
                    style = MaterialTheme.typography.labelLarge,
                    color = MaterialTheme.colorScheme.onSurface,
                )
            }
        }
    }
}

private const val MIN_LOGS = 3

private data class Snapshot(val condition: String, val hits: Int, val total: Int)

private fun snapshot(logs: List<ReliefLog>, saved: List<String>): Snapshot? {
    if (logs.size < MIN_LOGS) return null
    val helpful = logs.filter { it.rating >= 4 }
    val hits = helpful.size
    val total = logs.size
    val condition = bestCondition(helpful, saved) ?: return null
    return Snapshot(condition, hits, total)
}

private fun bestCondition(
    helpful: List<ReliefLog>,
    saved: List<String>,
): String? {
    val candidates = helpful.flatMap { it.notes.split(",", ";").map(String::trim) }
    val savedLower = saved.map { it.lowercase() }
    val match = savedLower.firstOrNull { saved ->
        candidates.any { it.lowercase().contains(saved) }
    }
    if (match != null) return saved.first { it.lowercase() == match }
    return helpful.firstOrNull()?.notes?.split(",", ";")?.firstOrNull()?.trim()
}

private fun labelFor(snap: Snapshot): String {
    val head = "Helpful for ${snap.condition}"
    val tail = "${snap.hits} of ${snap.total} sessions hit the mark"
    return "$head · $tail"
}

File: ios/StrainEase/Strain/StrainPersonalInsight.swift
import SwiftUI

/// Patient-specific "works for X" chip on the strain detail page.
///
/// Mirrors `src/components/saved/StrainPersonalInsight.tsx` from PR-W1:
/// shows once the patient has logged this strain at least 3 times, and
/// calls out the saved-condition match (e.g. "Helpful for insomnia · 3 of
/// 5 sessions hit the mark"). Tries each saved ailment first, falls back
/// to any-condition. Stays silent when there isn't enough data.
struct StrainPersonalInsight: View {
    let logs: [ReliefLog]
    var savedConditions: [String]

    /// 3 logs is the PR-W1 web threshold. Surface only when met.
    private let minLogs = 3

    private var snapshot: Snapshot? {
        guard logs.count >= minLogs else { return nil }
        let helpful = logs.filter { $0.fit == .justRight && $0.relief >= 4 }
        let hits = helpful.count
        let total = logs.count
        let condition = bestCondition(from: helpful)
        guard let condition else { return nil }
        return Snapshot(condition: condition, hits: hits, total: total)
    }

    var body: some View {
        if let snap = snapshot {
            HStack(spacing: 8) {
                Image(systemName: "chart.line.uptrend.xyaxis")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundStyle(Palette.primary)
                Text(label(for: snap))
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Palette.foreground)
                Spacer(minLength: 0)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 10)
            .background(Palette.primary.opacity(0.08), in: Capsule())
            .overlay(
                Capsule().strokeBorder(Palette.primary.opacity(0.25), lineWidth: 1)
            )
        }
    }

    private func label(for snap: Snapshot) -> String {
        // Same copy shape as PR-W1: "Helpful for X · N of M sessions hit the mark".
        let head = "Helpful for \(snap.condition)"
        let tail = "\(snap.hits) of \(snap.total) sessions hit the mark"
        return "\(head) · \(tail)"
    }

    /// Try saved conditions first, then any condition present in logs.
    private func bestCondition(from helpful: [ReliefLog]) -> String? {
        let candidates = helpful.flatMap { $0.conditions.map { $0.lowercased() } }
        let saved = savedConditions.map { $0.lowercased() }
        if let match = saved.first(where: candidates.contains) {
            return savedConditions.first(where: { $0.lowercased() == match }) ?? match
        }
        return helpful.flatMap { $0.conditions }.first
    }

    private struct Snapshot {
        let condition: String
        let hits: Int
        let total: Int
    }
}

#Preview("Hit-rate chip") {
    let sampleLogs = [
        ReliefLog.sampleSleep,
        ReliefLog.sampleSleep,
        ReliefLog.sampleSleep,
        ReliefLog.sampleSleep,
        ReliefLog.sampleSleep,
    ]
    return StrainPersonalInsight(logs: sampleLogs, savedConditions: ["Insomnia"])
        .padding()
}
